emcc.py

```diff
@@ -2937,62 +2918,54 @@ def in_directory(root, child):
-      return text[1:-1]
-    else:
-      return text
-
-  def parse_string_list_members(text, sep):
-    values = text.split(sep)
-    result = []
-    index = 0
-    while True:
-      current = values[index].lstrip() # Cannot safely rstrip for cases like: "HERE-> ,"
-      assert len(current), "string array should not contain an empty value"
-      first = current[0]
-      if not(first == "'" or first == '"'):
-        result.append(current.rstrip())
-      else:
-        start = index
-        while True: # Continue until closing quote found
-          assert index < len(values), "unclosed quoted string. expected final character to be '%s' in '%s'" % (first, values[start])
-          new = values[index].rstrip()
-          if not len(new) == 0 and new[-1] == first:
-            if start == index:
-              result.append(current[1:-1])
-            else:
-              result.append((current + sep + new)[1:-1])
-            break
-          else:
-            current += sep + values[index]
-            index += 1
+def evaluate_settings(settings):
+  # Handle aliases in settings flags. These are settings whose name
+  # has changed.
+  settings_aliases = {
+      'BINARYEN': 'WASM',
+      'BINARYEN_MEM_MAX': 'WASM_MEM_MAX',
+      # TODO: change most (all?) other BINARYEN* names to WASM*
+  }
 
-      index += 1
-      if index >= len(values):
-        break
-    return result
-
-  if text[0] == '[':
-    text = text.rstrip()
-    assert text[-1] == ']', 'unclosed opened string list. expected final character to be "]" in "%s"' % (text)
-    inner = text[1:-1]
-    if inner.strip() == "":
-      return []
-    else:
-      return parse_string_list_members(inner, ",")
-  else:
+  res = {}
+  for key, value in settings.items():
     try:
-      return int(text)
-    except ValueError:
-      return parse_string_value(text)
+      if key in settings_aliases:
+        key = settings_aliases[key]
+
+      deferred = False
+
+      if key == 'EXPORTED_FUNCTIONS':
+        res['EXPORTED_FUNCTIONS_RESPONSE_FILE'] = ''
+
+      if type(value) is str and value[0] == '@':
+        if key == 'EXPORTED_FUNCTIONS':
+          res['EXPORTED_FUNCTIONS_RESPONSE_FILE'] = value
+        if key in DEFERRED_RESPONSE_FILES:
+          deferred = True
+        else:
+          filename = value[1:]
+          with open(filename) as f:
+            try:
+              value = settings_parser(f).parse_list()
+            except Exception as e:
+              exit_with_error('error while parsing respose file @"%s": %s', filename, str(e))
+
+      if not deferred:
+        try:
+          default_type = type(getattr(shared.Settings, key))
+          if default_type != type(value):
+            if default_type is int:
+              value = shared.expand_byte_size_suffixes(value[0])
+            if default_type is str:
+              value = value[0]
+        except AttributeError:
+          pass
+
+      res[key] = value
+    except Exception as e:
```

Comment:
I'd love to avoid these kind of generic catches if possible.   Can't we verify the stuff we are parsing and give more accurate errors?